File: src/main/java/aoc2015/Day4.java
package aoc2015;

import common.Util;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class Day4
{
    private static long findHashWhichStartsWith( String input, String start ) throws NoSuchAlgorithmException
    {
        MessageDigest md = MessageDigest.getInstance("MD5");
        long i=1;
        while( true )
        {
            md.update((input + i++).getBytes());
            byte[] digest = md.digest();
            String hash = Util.bytesToHex( digest );
            if( hash.startsWith(start))
                break;
        }
        return i-1;
    }

    public static void main(String[] args) throws NoSuchAlgorithmException
    {
        assert findHashWhichStartsWith("abcdef", "00000") == 609043L;

        // part 1
        System.out.println( findHashWhichStartsWith("yzbqklnj", "00000"));

        //part 2
        System.out.println( findHashWhichStartsWith("yzbqklnj", "000000"));
    }
}

File: src/main/kotlin/aoc2015/Day4.kt
package aoc2015

import Solution
import java.security.MessageDigest

class Day4 : Solution {
    override fun part1(input: String): Any {
        return findMD5withPrefix(input, "00000")
    }

    override fun part2(input: String): Any {
        return findMD5withPrefix(input, "000000")
    }

    private fun findMD5withPrefix(key: String, prefix: String): Int {
        var counter = 0
        val md = MessageDigest.getInstance("MD5")
        while(true) {
            if (md5("$key$counter", md).startsWith(prefix)) {
                return counter
            }
            counter++
        }
    }

    @OptIn(ExperimentalStdlibApi::class)
    private fun md5(input: String, md: MessageDigest): String {
        return md.digest(input.toByteArray()).toHexString()
    }
}